### ISSUE TITLE
I know it's a bad idea option... (5.6 replica, to 5.7 master)

### ISSUE THREAD

**Issue by tomkrouper** [openark/orchestrator#264]
Is it possible to create an override option for when you are :100: aware of how bad the idea is?

```shell
$ orchestrator -c move-below -i hostname-56 -d hostname-57
14:04:18 FATAL instance hostname-56:3306 has version 5.6.35-log, which is lower than 5.7.18-log on hostname-57:3306
```

We are upgrading a cluster to 5.7 but want to keep a 5.6 replica "just in case". To move it under the new master is impossible with `orchestrator`. Is there a way we could add a feature to say, "no, really, do it anyhow?"

**Comment by sjmudd** [comment]
I've seen the need for the same functionality and think it would be a good idea. So probably we just need a --force flag and adjust the code to not generate a fatal message if the checks fail and you provide that --force option.

**Comment by shlomi-noach** [comment]
What other operations may be affected by a `--force` or a `--skip-sanity-checks` flag?

**Comment by sjmudd** [comment]
I have seen this come up a couple of times but don't remember when. I'll open an issue should I see the need for a --force flag again.

As for --move-below I'd think that --relocate and similar commands where such checks exist probably need the same potential --force override.

**Comment by shlomi-noach** [comment]
yes, all Relocate, MatchBelow, MoveBelow, MoveUp etc. functions all share the same kind of test and will all have to be refactored. I somehow have in the back of my mind another use case that doesn't have to do with relocation (nor recovery). Perhaps coffee is in place.

**Comment by shlomi-noach** [comment]
This is somewhat more complex than originally considered, for programmatic reasons.

Typically, I would set some `IgnoreSanityChecks` boolean in `CLIFlags` via `--ignore-sanity-checks` and would check that value before considering calling `CanReplicate()`.

However, with API calls that is different: if this is to be supported via API calls as well (and via `orchestrator-client`) then this can't be just a global variable; each incoming request may treat an operation differently: with or without checks.

Thus, we need to introduce an `OperationContext`. So that a function like `MoveBelow()` would get such an argument and know how to react.

However, there are many paths leading to `MoveBelow()`, to `MatchBelow()`, to `MoveBelowGTID()` etc, some of which come from automated failovers, some from general purpose multi-instance ops etc. So each of those will now need to get a `OperationContext` as well. This is going to change the signature of dozens of functions and so I need to consider how to make this change as elegant as possible.

Once again, the lack of (thread) context in `golang` turns to be one of the language's most obvious flaws.

**Comment by pboros** [comment]
@shlomi-noach For this use case (upgrading) Wouldn't it be enough to add a switch to ignore the version check (every other check should be fine).

https://github.com/github/orchestrator/blob/760bf40ba11b6359773353f0fae03fdb81c8ad3a/go/inst/instance.go#L365

And put it under a variable name IgnoreDifferentVersionsForUpgrade or something, which turns off jsut this check. it would be a lot less risky than ignoring all the sanity checks. If you agree with this approach, I will submit a PR for this one.

**Comment by tomkrouper** [comment]
@pboros I know @shlomi-noach is not in this week and I know we all have been busy with other priorities. This got a lower priority because my initial reason for opening the issue was worked around and/or went away. I would definitely suggest doing the PR. If it works for you and it would have worked for me, it sounds like a great (and when documented a properly known dangerous) feature that others could use. Thanks for offering to look at it.

**Comment by rwultsch** [comment]
@shlomi-noach would you look favorably on a patch which set created a new flag in the configuration file for the Orchestrator server to always skip version checks?

**Comment by shlomi-noach** [comment]
@rwultsch that sounds fine; please name the flag `IgnoreCheck*` or `SkipCheck*` so that the name describes the risk involved.
